### PR TITLE
[16.0][MIG] web_widget_x2many_2d_matrix

### PR DESCRIPTION
**Pull request by tarteo**

### CODE REVIEW

**Comment by tarteo** [comment]
I'll also migrate the timesheet sheet module (hr_timesheet_sheet).

File: web_widget_x2many_2d_matrix/__manifest__.py [hunk 1]
```diff
@@ -24,11 +25,14 @@
     "installable": True,
     "assets": {
         "web.assets_backend": [
-            "web_widget_x2many_2d_matrix/static/src/scss/web_widget_x2many_2d_matrix.scss",
```
Comment (astirpe) [inline web_widget_x2many_2d_matrix/__manifest__.py]: There's a committed file `x2many_2d_matrix_field.scss` that is empty, maybe a leftover from development? Could be safely removed IMO.

Also you could use the wildcard here to make the "web.assets_backend" part more light.

**Review by astirpe — CHANGES_REQUESTED** [review]
@tarteo thank you for this great work! The code looks very clean now.

We are all eager to test it with the timesheet sheet module :wink:

**Comment by legalsylvain** [comment]
Hi @Tarteo. Thanks for porting this usefull module.
could you take a look on [open PRs](https://github.com/OCA/web/pulls?q=is%3Apr+is%3Aopen+matrix+) that bring some fixes or improvments ? and cherry pick the interesting commits ? specially on V15 modules : 

- https://github.com/OCA/web/pull/2380 by @vincent-hatakeyama
- https://github.com/OCA/web/pull/2379 by @vincent-hatakeyama
- https://github.com/OCA/web/pull/2361 by @vincent-hatakeyama

Thanks !

File: web_widget_x2many_2d_matrix/__manifest__.py (hunk 1 above)
Comment (tarteo) [reply web_widget_x2many_2d_matrix/__manifest__.py]: Forgot to migrate migrate that file, thanks. It's fixed now.

**Review by astirpe — APPROVED** [review]
Code review: LGTM!
Functional test: by testing the timesheet sheet widget of https://github.com/OCA/timesheet/pull/580 I implicitly tested this matrix module and it works properly
Thanks!

**Comment by tarteo** [comment]
@legalsylvain I'm currently working on something else so I don't have time to implement extra features now. I think some PRs fixing things are already fixed by the migration. Can you please review the migation as is. Thanks! :smile:

File: web_widget_x2many_2d_matrix/static/src/components/x2many_2d_matrix_field/x2many_2d_matrix_field.xml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<templates>
+    <t t-name="web_widget_x2many_2d_matrix.X2Many2DMatrixField" owl="1">
+        <div class="table-responsive">
+            <X2Many2DMatrixRenderer
+                list="list"
+                matrixFields="props.matrixFields"
+                showRowTotals="props.showRowTotals"
+                showColumnTotals="props.showColumnTotals"
+                setDirty="props.setDirty"
```
Comment (sonhd91) [inline web_widget_x2many_2d_matrix/static/src/components/x2many_2d_matrix_field/x2many_2d_matrix_field.xml]:
```suggestion
                setDirty="(isDirty) => this.setDirty(isDirty)"
```
fixed this error
```
`UncaughtPromiseError > OwlError
Uncaught Promise > Invalid props for component 'X2Many2DMatrixRenderer': 'setDirty' is undefined (should be a function)
OwlError: Invalid props for component 'X2Many2DMatrixRenderer': 'setDirty' is undefined (should be a function)
    at Object.validateProps (http://localhost:8069/web/assets/debug/web.assets_common.js:19239:19) (/web/static/lib/owl/owl.js:3069)
    at X2Many2DMatrixField.template (eval at compile (http://localhost:8069/web/assets/debug/web.assets_common.js:21606:16), <anonymous>:11:13) (/web/static/lib/owl/owl.js:5436)
    at Fiber._render (http://localhost:8069/web/assets/debug/web.assets_common.js:17791:38) (/web/static/lib/owl/owl.js:1621)
    at Fiber.render (http://localhost:8069/web/assets/debug/web.assets_common.js:17783:18) (/web/static/lib/owl/owl.js:1613)
    at ComponentNode.initiateRender (http://localhost:8069/web/assets/debug/web.assets_common.js:18505:23) (/web/static/lib/owl/owl.js:2335)`
``` 

Reproduce
1. Create wizard like in README
2. Add action to menu-action open the wizard form, error showup

**Comment by Saran440** [comment]
Hi,

In v16, It error
Selection_011


with my code
```
<field
    name="line_ids"
    widget="x2many_2d_matrix"
    field_x_axis="date_range_id"
    field_y_axis="name"
    field_value="amount"
>
    <tree>
        <field name="date_range_id" />
        <field name="name" />
        <field name="amount" />
    </tree>
</field>
```

result in v15
Selection_012

**Comment by PieterPaulussen** [comment]
Code review: Broken
I've tested the module thoroughly and support for a Many2one value field is missing. I've created a PR to address the issue [here](https://github.com/tarteo/web/pull/13), although I personally think it still falls short of the features it should have.

Features that need be restored/added/updated:
- Add possibility to specify the aggregation field type (int, float, ...). Current code version uses the same field type to display the aggregation value. There is also no check to validate if the row or column CAN be aggregated at all (which seemed to be present in older versions?)
- The use of the word `list` seems confusing. I've refactored the use of list to `matrixRows` which seems more appropriate and clarifies its purpose.

File: web_widget_x2many_2d_matrix/static/src/components/x2many_2d_matrix_field/x2many_2d_matrix_field.xml (hunk 2 above)
Comment (tarteo) [reply web_widget_x2many_2d_matrix/static/src/components/x2many_2d_matrix_field/x2many_2d_matrix_field.xml]: Thanks! I'll fix it.

**Comment by tarteo** [comment]
@PieterPaulussen You realize I did a complete rewrite of this field widget? 

> Add possibility to specify the aggregation field type (int, float, ...). Current code version uses the same field type to display the aggregation value. There is also no check to validate if the row or column CAN be aggregated at all (which seemed to be present in older versions?)

This is certainly not the case in previous versions. You can't differ from the aggregated field type. Checking if the field can be aggregated is literally one line of code, please calm down...

> The use of the word list seems confusing. I've refactored the use of list to matrixRows which seems more appropriate and clarifies its purpose.

list is not the matrix it's a list of records. there's a different variable for it called "matrix" in X2Many2DMatrixRenderer. Lists are different from matrices.

**Comment by tarteo** [comment]
@Saran440 Thanks I'll look into it.

File: web_widget_x2many_2d_matrix/static/src/components/x2many_2d_matrix_field/x2many_2d_matrix_field.xml (hunk 2 above)
Comment (tarteo) [reply web_widget_x2many_2d_matrix/static/src/components/x2many_2d_matrix_field/x2many_2d_matrix_field.xml]: Should be fixed now

**Comment by tarteo** [comment]
Hey @Saran440 It should be fixed now

**Review by Saran440 — APPROVED** [review]
Functional Test :+1:

File: web_widget_x2many_2d_matrix/static/src/components/x2many_2d_matrix_field/x2many_2d_matrix_field.xml (hunk 2 above)
Comment (astirpe) [reply web_widget_x2many_2d_matrix/static/src/components/x2many_2d_matrix_field/x2many_2d_matrix_field.xml]: @sonhd91 is it ok for you now? Can you approve this PR?

**Comment by ghost** [comment]
Functional test with web_widget_x2many_2d_matrix_example: 👍 for simple field. 👎 for selection and many2one fields.

**Comment by tarteo** [comment]
@appstogrow Thanks I think I can look at it this week. Using your migration of the example module, thanks! I only tested it with the timesheet sheet module.

**Review by ernestotejeda — APPROVED** [review]
Tested on web_widget_x2many_2d_matrix_example https://github.com/OCA/web/pull/2520 👍🏼

**Comment by pedrobaeza** [comment]
@tarteo are you going to check the m2o case?

**Comment by tarteo** [comment]
@pedrobaeza yes

**Comment by pedrobaeza** [comment]
Great, thanks.

/ocabot merge nobump

**Comment by MiquelRForgeFlow** [comment]
Merge should be stopped and restarted. @pedrobaeza

**Comment by MiquelRForgeFlow** [comment]
@sbidoul we should implement an option of /ocabot merge stop for these cases

**Comment by MiquelRForgeFlow** [comment]
@tarteo a new PR should be done to add your last commit!

**Comment by tarteo** [comment]
Just pushed a fix for many2one functionality

**Pull request merged**